File: minWordDistance2.java
// Time complexity = O(n)
// Space complexity = O(n), we use n for the map but its a 1 time thing during the initialization, also used n for the lists
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
// We maintain a hashMap of all the words and the list of its indexes
class WordDistance {
    HashMap<String, List<Integer>> map;

    // itterate over the words and fill the map
    public WordDistance(String[] wordsDict) {
        map = new HashMap<>();
        int i=0;

        for (String word: wordsDict) {
            if(!map.containsKey(word)) {
                map.put(word, new ArrayList<>());
            }
            map.get(word).add(i);
            i++;
        }
    }

    // we get the list of indices for both the words and we check for the min distance between them by incrementing the smaller index so that the difference can be minimized
    public int shortest(String word1, String word2) {
        List<Integer> l1 = map.get(word1);
        List<Integer> l2 = map.get(word2);
        int min = Integer.MAX_VALUE;

        int idx1 = 0;
        int idx2 = 0;

        while (idx1 < l1.size() && idx2 < l2.size()) {
            min = Math.min(min, Math.abs(l1.get(idx1)-l2.get(idx2)));
            if (l1.get(idx1) < l2.get(idx2)) idx1++;
            else idx2++;
        }
        return min;
    }
}

File: Min_word_Distance2.py
# Time complexity : O(p + q) --> occurences of word1 and word2
# Space complexity : O(1) --> space of hashmap is in constructor
# Leetcode : Solved and submitted

class WordDistance:
    def __init__(self, wordsDict: List[str]):
      # declare a hashmap for the word and it's occurences
        self.hashmap = {}
        
        # make a hashmap of words as key and it's occurence index as the value in an array list
        for i, word in enumerate(wordsDict):
            if word not in self.hashmap:
                self.hashmap[word] = []
            self.hashmap[word].append(i)
        
    def shortest(self, word1: str, word2: str) -> int:
        # fetch the list of both words
        lis1 = self.hashmap[word1]
        lis2 = self.hashmap[word2]
        
        p1, p2, min_val = 0, 0, float('inf')
        
        # traverse until one of the list is finished
        while p1 < len(lis1) and p2 < len(lis2):
          
            # find the min value
            min_val = min(min_val, abs(lis1[p1] - lis2[p2]))
            
            # update the min value from the list of lis1 and lis2
            if lis1[p1] < lis2[p2]:
                p1 += 1
            else:
                p2 += 1
                
        # return the min dist and final value
        return min_val


# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)
